flink-connectors/flink-connector-pubsub/src/main/java/org/apache/flink/streaming/connectors/pubsub/PubSubSink.java

```diff
@@ -0,0 +1,239 @@
+import java.util.concurrent.TimeUnit;
+
+import static com.google.cloud.pubsub.v1.SubscriptionAdminSettings.defaultCredentialsProviderBuilder;
+
+/**
+ * A sink function that outputs to PubSub.
+ *
+ * @param <IN> type of PubSubSink messages to write
+ */
+public class PubSubSink<IN> extends RichSinkFunction<IN> {
+	private static final Logger LOG = LoggerFactory.getLogger(PubSubSink.class);
+
+	private final Credentials credentials;
+	private final SerializationSchema<IN> serializationSchema;
+	private final String projectName;
+	private final String topicName;
+	private final String hostAndPort;
+
+	private transient Publisher publisher;
+
+	private PubSubSink(
+		Credentials credentials,
+		SerializationSchema<IN> serializationSchema,
+		String projectName,
+		String topicName,
+		String hostAndPort) {
+		this.credentials = credentials;
+		this.serializationSchema = serializationSchema;
+		this.projectName = projectName;
+		this.topicName = topicName;
+		this.hostAndPort = hostAndPort;
+	}
+
+	private transient ManagedChannel managedChannel = null;
+	private transient TransportChannel channel = null;
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		Publisher.Builder builder = Publisher
+			.newBuilder(ProjectTopicName.of(projectName, topicName))
+			.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+
+		if (hostAndPort != null) {
+			managedChannel = ManagedChannelBuilder
+				.forTarget(hostAndPort)
+				.usePlaintext(true) // This is 'Ok' because this is ONLY used for testing.
+				.build();
+			channel = GrpcTransportChannel.newBuilder().setManagedChannel(managedChannel).build();
+			builder.setChannelProvider(FixedTransportChannelProvider.create(channel));
+		}
+
+		publisher = builder.build();
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		shutdownPublisher();
+		shutdownTransportChannel();
+		shutdownManagedChannel();
+	}
+
+	private void shutdownPublisher() {
+		try {
+			publisher.shutdown();
+		} catch (Exception e) {
+			LOG.info("Shutting down Publisher failed: {}", e.getMessage());
+		}
+	}
+
+	private void shutdownTransportChannel() {
+		if (channel == null) {
+			return;
+		}
+		try {
+			channel.close();
+		} catch (Exception e) {
+			LOG.info("Shutting down TransportChannel failed: {}", e.getMessage());
+		}
+	}
+
+	private void shutdownManagedChannel() {
+		if (managedChannel == null) {
+			return;
+		}
+		try {
+			managedChannel.shutdownNow();
+			managedChannel.awaitTermination(1000L, TimeUnit.MILLISECONDS);
+		} catch (Exception e) {
+			LOG.info("Shutting down ManagedChannel failed: {}", e.getMessage());
+		}
+	}
+
+	@Override
+	public void invoke(IN message, SinkFunction.Context context) {
+		PubsubMessage pubsubMessage = PubsubMessage
+			.newBuilder()
+			.setData(ByteString.copyFrom(serializationSchema.serialize(message)))
+			.build();
+		publisher.publish(pubsubMessage);
+		publisher.publishAllOutstanding();
```

Comment:
My rationale was that you do not want PubSub to batch / buffer messages. Because a checkpoint may be marked as successful but the PubSub messages have not been sent yet and on a failure you could have PubSub messages that won't ever be sent because from Flink's perspective they have already been send.

I think an improvement of this sink could be to add a buffer in Flink state, so we would call the publishAllOutstanding() once every n messages. Do you have any ideas how to approach something like that?

Reply:
From what I see in the `Publisher`, it looks like `publishAllOutstanding` is a "force flush" option, which overrules the more relaxed scheduling done in `publish`. Calling `publish` does not seem necessary, if you do not want any buffering to happen. This, however, can induce the typical overhead which buffering aims to mitigate. 

I think the closest example to what you want to achieve here is `FlinkKafkaProducer09` and it's base class `FlinkKafkaProducerBase`. It provides at-least-once guarantees. Notice how it uses  `snapshotState` method from `CheckpointedFunction` to get notified about the checkpoint requests and flushes all pending messages prior to taking a snapshot. This allows to make sure that even in case of a failure messages are sent out at least once. Kafka's `producer.flush()` is semantically equivalent to `publishAllOutstanding()` from PubSub. 

As a side note: implementing exactly-once would be a much more sophisticated endeavor, requiring implementation of two-phase commit protocol + support of transactional capabilities from the target system (I am having trouble finding such details about PubSub), see `FlinkKafkaProducer011` if you are curious.
 @rmetzger please fact check my description.

Reply:
@afedulov Yeah this makes a lot of sense! I've added it to the PubSubSink.

Exactly once won't ever be available I think as PubSub itself is at least-once. https://cloud.google.com/pubsub/docs/subscriber#at-least-once-delivery We've seen messages being resent multiple times days after they had been delivered under some scenarios.. :)